File: TRPTSTIC.py
for _ in range(int(input())):
	n, m, k = map(int, input().split())
	grid = [ [0 for i in range(m+1) ] ]
	for i in range(1, n+1): grid.append([0] + list(map(int, input().split())))
	
	pref = [ [0 for i in range(m+1)] for j in range(n+1)]
	for i in range(1, n+1):
		for j in range(1, m+1):
			pref[i][j] = pref[i-1][j] + pref[i][j-1] - pref[i-1][j-1] + grid[i][j]
	if pref[n][m] <= k:
		print(-1)
		continue
	
	def getsum(l1, r1, l2, r2):
		return pref[l2][r2] - pref[l1-1][r2] - pref[l2][r1-1] + pref[l1-1][r1-1]
	
	lo, hi = -1, n+m+1
	while lo+1 < hi:
		mid = (lo + hi)//2
		mxsum = 0
		for i in range(1, n+1):
			for j in range(1, m+1):
				if grid[i][j] == 0: continue
				l1 = max(1, i-mid)
				l2 = min(n, i+mid)
				r1 = max(1, j-mid)
				r2 = min(m, j+mid)
				mxsum = max(mxsum, getsum(l1, r1, l2, r2))
		if mxsum <= k: lo = mid
		else: hi = mid
	print(hi)

File: TRPTSTIC.cpp
#define ll long long int
#include<bits/stdc++.h>
#define loop(i,a,b) for(ll i=a;i<b;++i)
#define rloop(i,a,b) for(ll i=a;i>=b;i--)
#define in(a,n) for(ll i=0;i<n;++i) cin>>a[i];
#define pb push_back
#define mk make_pair
#define all(v) v.begin(),v.end()
#define dis(v) for(auto i:v)cout<<i<<" ";cout<<endl;
#define display(arr,n) for(int i=0; i<n; i++)cout<<arr[i]<<" ";cout<<endl;
#define fast ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);srand(time(NULL));
#define l(a) a.length()
#define s(a) (ll)a.size()
#define fr first
#define sc second
#define mod 1000000007
#define endl '\n'
#define yes cout<<"Yes"<<endl;
#define no cout<<"No"<<endl;
using namespace std;
#define debug(x) cerr << #x<<" "; _print(x); cerr << endl;
void _print(ll t) {cerr << t;}
void _print(int t) {cerr << t;}
void _print(string t) {cerr << t;}
void _print(char t) {cerr << t;}
void _print(double t) {cerr << t;}
template <class T, class V> void _print(pair <T, V> p);
template <class T> void _print(vector <T> v);
template <class T> void _print(set <T> v);
template <class T, class V> void _print(map <T, V> v);
template <class T> void _print(multiset <T> v);
template <class T, class V> void _print(pair <T, V> p) {cerr << "{"; _print(p.fr); cerr << ","; _print(p.sc); cerr << "}";}
template <class T> void _print(vector <T> v) {cerr << "[ "; for (T i : v) {_print(i); cerr << " ";} cerr << "]";}
template <class T> void _print(set <T> v) {cerr << "[ "; for (T i : v) {_print(i); cerr << " ";} cerr << "]";}
template <class T> void _print(multiset <T> v) {cerr << "[ "; for (T i : v) {_print(i); cerr << " ";} cerr << "]";}
template <class T, class V> void _print(map <T, V> v) {cerr << "[ "; for (auto i : v) {_print(i); cerr << " ";} cerr << "]";}

ll add(ll x,ll y)  {ll ans = x+y; return (ans>=mod ? ans - mod : ans);}
ll sub(ll x,ll y)  {ll ans = x-y; return (ans<0 ? ans + mod : ans);}
ll mul(ll x,ll y)  {ll ans = x*y; return (ans>=mod ? ans % mod : ans);}


void solve(){
    ll n,m,k;   cin>>n>>m>>k;
    vector<vector<ll>> v(n,vector<ll>(m));
    loop(i,0,n) loop(j,0,m) cin>>v[i][j];
    vector<vector<ll>> vec = v;
    loop(i,1,n) v[i][0]+=v[i-1][0];
    loop(j,1,m) v[0][j]+=v[0][j-1];
    loop(i,1,n) loop(j,1,m) v[i][j]+=v[i-1][j]+v[i][j-1]-v[i-1][j-1];
    ll ans = INT_MAX;
    loop(i,0,n){
        loop(j,0,m){
            if(!vec[i][j])  continue;
            ll l = 0,r = max(n,m);
            while(l<=r){
                ll mid = (l+r)/2;
                ll x = min(n-1,i+mid),y = min(m-1,j+mid);
                ll sum = v[x][y];
                if(i-mid>0) sum-=v[i-mid-1][y];
                if(j-mid>0) sum-=v[x][j-mid-1];
                if(i-mid>0 && j-mid > 0)    sum+=v[i-mid-1][j-mid-1];
                if(sum>=k+1)    r = mid-1,ans = min(ans,mid);
                else l = mid+1;
            }
        }
    }
    if(ans == INT_MAX)  ans = -1;
    cout<<ans<<endl;
}


int main()
{
    fast
    int t; cin>>t;
    
    while(t--) solve();
    return 0;
}